File: backend_bundle_staging_tmp/server/middleware/rateLimit.js
const { env } = require('../config/env');

const rateLimitEnabled = env.rateLimit?.enabled !== false;

const WINDOW_SECONDS = env.rateLimit?.windowSeconds || 60;
const DEFAULT_LIMIT = env.rateLimit?.maxRequests || 300;
const EXPENSIVE_LIMIT = env.rateLimit?.maxRequestsExpensive || 80;
const AUTH_LIMIT = env.rateLimit?.maxRequestsAuth || 40;

const hits = new Map();

const getClientIp = (req) => {
  const candidate = req.headers['cf-connecting-ip']
    || req.headers['x-forwarded-for']
    || req.socket?.remoteAddress
    || req.ip
    || 'unknown';
  if (Array.isArray(candidate)) {
    return String(candidate[0] || 'unknown').split(',')[0].trim() || 'unknown';
  }
  return String(candidate).split(',')[0].trim() || 'unknown';
};

const isApiPath = (path) => (path || '').startsWith('/api');

const isExempt = (path) => {
  if (!path) return true;
  if (path === '/api/health' || path === '/api/help') return true;
  if (path === '/api/payments/stripe/webhook') return true;
  return false;
};

const bucketFor = (path) => {
  if (!path) return 'default';
  if (path.startsWith('/api/auth')) return 'auth';
  if (path.startsWith('/api/orders/sales-rep')) return 'orders_sales_rep';
  if (path.startsWith('/api/woo')) return 'woo';
  if (path.startsWith('/api/shipping')) return 'shipping';
  return 'default';
};

const limitForBucket = (bucket) => {
  switch (bucket) {
    case 'auth':
      return AUTH_LIMIT;
    case 'orders_sales_rep':
    case 'woo':
    case 'shipping':
      return EXPENSIVE_LIMIT;
    default:
      return DEFAULT_LIMIT;
  }
};

const pruneBucket = (timestamps, cutoff) => {
  while (timestamps.length > 0 && timestamps[0] < cutoff) {
    timestamps.shift();
  }
};

const rateLimit = (req, res, next) => {
  if (!rateLimitEnabled) {
    return next();
  }

  if (req.method === 'OPTIONS') {
    return next();
  }

  const path = req.path || req.originalUrl || '';
  if (!isApiPath(path) || isExempt(path)) {
    return next();
  }

  const ip = getClientIp(req);
  const bucket = bucketFor(path);
  const limit = limitForBucket(bucket);
  const now = Date.now();
  const cutoff = now - WINDOW_SECONDS * 1000;
  const key = `${ip}:${bucket}`;

  const timestamps = hits.get(key) || [];
  pruneBucket(timestamps, cutoff);

  if (timestamps.length >= limit) {
    const oldest = timestamps[0] || now;
    const retryAfterSeconds = Math.max(1, Math.ceil((WINDOW_SECONDS * 1000 - (now - oldest)) / 1000));
    res.setHeader('Retry-After', String(retryAfterSeconds));
    return res.status(429).json({
      error: 'Too many requests. Please wait a moment and try again.',
      code: 'RATE_LIMITED',
      retryAfterSeconds,
    });
  }

  timestamps.push(now);
  hits.set(key, timestamps);
  return next();
};

module.exports = { rateLimit };

File: tmp/backend_v1.9.45/python_backend/middleware/rate_limit.py
from __future__ import annotations

import os
import threading
import time
from collections import deque
from typing import Deque, Dict, Tuple

from flask import Flask, jsonify, request


def _truthy(value: str) -> bool:
    return value.strip().lower() in ("1", "true", "yes", "on")


_RATE_LIMIT_ENABLED = _truthy(os.environ.get("RATE_LIMIT_ENABLED", "true"))
_WINDOW_SECONDS = int(os.environ.get("RATE_LIMIT_WINDOW_SECONDS", "60").strip() or 60)
_WINDOW_SECONDS = max(10, min(_WINDOW_SECONDS, 10 * 60))

_DEFAULT_MAX = int(os.environ.get("RATE_LIMIT_MAX_REQUESTS", "300").strip() or 300)
_DEFAULT_MAX = max(30, min(_DEFAULT_MAX, 5000))

_EXPENSIVE_MAX = int(os.environ.get("RATE_LIMIT_MAX_REQUESTS_EXPENSIVE", "80").strip() or 80)
_EXPENSIVE_MAX = max(10, min(_EXPENSIVE_MAX, 1000))

_lock = threading.Lock()
_hits: Dict[Tuple[str, str], Deque[float]] = {}


def _is_api_path(path: str) -> bool:
    return (path or "").startswith("/api")


def _is_exempt(path: str) -> bool:
    if not path:
        return True
    if path.startswith("/api/health"):
        return True
    return False


def _limit_for_path(path: str) -> int:
    path = path or ""
    if path.startswith("/api/orders/sales-rep") or path.startswith("/api/woo"):
        return _EXPENSIVE_MAX
    return _DEFAULT_MAX


def init_rate_limit(app: Flask) -> None:
    """
    Very small in-memory rate limiter to prevent runaway polling / thundering herds
    from exhausting low cPanel resource plans.
    """
    if not _RATE_LIMIT_ENABLED:
        return

    @app.before_request
    def _rate_limit() -> None:  # type: ignore[return-value]
        if request.method == "OPTIONS":
            return
        if not _is_api_path(request.path) or _is_exempt(request.path):
            return

        limit = _limit_for_path(request.path)
        ip = request.headers.get("CF-Connecting-IP") or request.headers.get("X-Forwarded-For") or request.remote_addr or "unknown"
        ip = ip.split(",")[0].strip() if ip else "unknown"

        key = (ip, request.path)
        now = time.time()
        cutoff = now - _WINDOW_SECONDS

        with _lock:
            bucket = _hits.get(key)
            if bucket is None:
                bucket = deque()
                _hits[key] = bucket

            while bucket and bucket[0] < cutoff:
                bucket.popleft()

            if len(bucket) >= limit:
                retry_after = max(1, int(_WINDOW_SECONDS - (now - bucket[0])) if bucket else _WINDOW_SECONDS)
                response = jsonify(
                    {
                        "error": "RATE_LIMITED",
                        "message": "Too many requests. Please wait a moment and try again.",
                        "retryAfterSeconds": retry_after,
                    }
                )
                response.status_code = 429
                response.headers["Retry-After"] = str(retry_after)
                return response

            bucket.append(now)

